[big merge][student-teacher] Sophiex/dev/ssl losses 1043

sophie-xhonneux opened the pull request:
## Description

[DRAFT] PR for introducing the losses for SSL student-teacher latent losses. This PR will rely on both the abstract loss calculator #1178  as well as the abstract target/aux class #1179

The idea is to get early feedback and notice issues my making code more concrete

## Issue Number

Closes #1043 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

wael-mika committed:
Implemented the first draft of the Cropping feature

sophie-xhonneux committed:
Abstract class for target/aux computation

Implemented Identity class

TODO: implement EMATeacher

sophie-xhonneux committed:
Abstract class for target/aux computation

Implemented Identity class

TODO: implement EMATeacher

sophie-xhonneux committed:
Start implementing the EMA Teacher

The big question on the EMA teacher side to me is how to allow for a
fleixble teacher and student architecture that can differ

We updated some APIs of the abstract base class to allow the ema_model
forward, subject to change given the loss calculator, which is imho the
second big question mark

sophie-xhonneux committed:
Start implementing the EMA Teacher

The big question on the EMA teacher side to me is how to allow for a
fleixble teacher and student architecture that can differ

We updated some APIs of the abstract base class to allow the ema_model
forward, subject to change given the loss calculator, which is imho the
second big question mark

Jubeku committed:
adding loss calculator base class

sophie-xhonneux committed:
Option for constructing teacher model flexibly

sophie-xhonneux committed:
Option for constructing teacher model flexibly

sophie-xhonneux committed:
Extract get batch size util function

Easier to read and as batchsize gets more complicated in SSL this will
be a useful abstraction

sophie-xhonneux committed:
Fix mismatched dtypes in the target computation

It runs so far. Next steps:
 - Route all the config options
 - Start writing the loss functions to understand the state requirements

sophie-xhonneux committed:
Fix mismatched dtypes in the target computation

It runs so far. Next steps:
 - Route all the config options
 - Start writing the loss functions to understand the state requirements

sophie-xhonneux committed:
Lay groundwork for SSL losses

This involves creating stateful classes for each of the losses and the
EMATeacher being able to run additional neural network heads for these
losses.

Jubeku committed:
abstract loss calc structure

sophie-xhonneux committed:
Add the SSL Loss Processing classes

sophie-xhonneux committed:
Write part of the TargetProcessing forward

TODO: create the various teacher head modules and run them.
TODO: merge the abstract loss calculator and create the SSL one

Jubeku committed:
add abstract method to loss calculator base class

Jubeku committed:
add latent loss class

clessig commented on src/weathergen/train/ssl_losses_utils.py:
```diff
@@ -0,0 +1,304 @@
+import torch.nn.functional as F
+
+
+def lossfunc(t, s, temp):
```
The name is not very descriptive :) Maybe `latent_logit_loss.py`? JEPA uses MAE (and one could conceivably replace by MSE) which are already implemented in loss.py. Ideally we could reuse what is there.

clessig reviewed (commented):
Didn't look through the actual computations line by line since it seems this copy-paste from the reference code?

sophie-xhonneux committed:
Add latent prediction heads to the Model

After much consideration I decided to add the latent prediction heads to
the Model, because they also need to benefit from exponential moving
average of the weights and this gets unnecessarily cumbersome if they
are outside the Model.

TODO: make JEPA different between student and teacher
TODO: use this new structure in EMATeacher

sophie-xhonneux committed:
Adapt forward function for latent prediction heads

To prevent crazy nesting of model output values we created a ModelOutput
Dataclass (akin to how it is done in huggingface), and we run all the
latent_prediction heads.

shmh40 committed:
rough first effort producing globaland local views

shmh40 committed:
update to return 6 tuple from iter in multi-stream-data-sampler, with locals_prepared

sophie-xhonneux committed:
Start piping configs through model, trainer, etc

Will need adapting based on the abstract loss calculator

Currently is awaiting the streams data branch to check piping of data
and configuring this

Jubeku committed:
update loss calc config and rename files

tjhunter commented on src/weathergen/train/trainer.py:
```diff
@@ -313,8 +315,19 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
+                rampup_ratio=cf.get("ema_ramp_up_ratio", 0.09),
+                is_model_sharded=(cf.with_ddp and cf.with_fsdp),
+            )
+        elif cf["training_mode"] == "student-teacher":
```
small comment: prefer in general `cf.get(...)` for backward compatibilty

tjhunter reviewed (commented):
some initial comments. looking forward to seeing it in action. 

High level comment: the current teacher-student framework wraps the whole model. Do we want that? I always thought it would be applied more locally up to the global assimilation engine. It would simplify future interactions with the diffusion part in the forecasting engine.

sophie-xhonneux commented on src/weathergen/train/ssl_losses_utils.py:
```diff
@@ -0,0 +1,304 @@
+
+class iBOTPatchTargetProcessing(nn.Module):
+    """
+    Code taken and adapted from the official DINOv2 implementation
```
Let's actually discuss that last point with Christian, we may want to avoid that?

sophie-xhonneux commented on src/weathergen/train/target_and_aux_ssl_teacher.py:
```diff
@@ -0,0 +1,78 @@
+def get_target_postprocessing(target_losses: list[str], **kwargs):
+    return_dict = {}
+    for loss_name in target_losses:
+        if loss_name == "iBOT":
```
when was this introduced? not sure how readable it is more people :/

tjhunter committed:
adding dinov2 notice

sophie-xhonneux committed:
Draft Student Teacher Loss Calculator

TODO: initialise it and register
TODO: weight the loss
TODO: route the kwargs
TODO: check shapes of tensors

Jubeku committed:
restructure loss modules

Jubeku committed:
add ModelOutput dataclass

sophie-xhonneux committed:
Use infra provided by Abstract Loss Calc

Completes config option routing, weighting, and registering TODOs

sophie-xhonneux committed:
Run Ruff

clessig committed:
NOT WORKING: initial draft for index-based masking. Implemented for random and healpix masking. Open issues with _coords_local, centroids and probably other things.

sophie-xhonneux committed:
Fix class being in the wrong file

sophie-xhonneux committed:
Ensure data pipes through model and target

This is a DRAFT!

This commit assumes that the data augmentations of the stream_data
objectsee shmh40/dev/global-local will fit into the Batch data class
(trainer.py).

The goal was to ensure all data reaches the LossCalculator.

Large scale todos:
- Pass Metadata about local2global correspondance, etc to the LossCalculator
- Upgrade the Model heads to produce the correct dimensions
- Verify the Data shapes against DINOv2

Smaller todos:
- Ensure teacher params are requires_grad = false
- clean up code

clessig committed:
NOT WORKING: Finished src, target still to be done.

clessig committed:
Masking target is working in principle but errors when feeding data to the model.

clessig committed:
Working version for ERA5, NPP-ATMS. Problems with SYNOP with empty cell handling

clessig committed:
Minor cleanup

clessig committed:
Fixed linting

clessig committed:
Fixed remaining problems that occured for NPP-ATMS and SYNOP.
TODO:
- Forecast still needs to be adapted
- Some more cleanup of variable naming, return values etc

clessig committed:
Enabled support for forecast. Cleaned up some bits and pieces.

Jubeku committed:
merge develop

Jubeku committed:
mv streams_data declaration under if condition

Jubeku committed:
add weight to loss config, add toy loss class LossPhysicalTwo

sophie-xhonneux committed:
Update Abstract Target class based on needs for SSL losses

clessig committed:
Removing centroids options for embedding that was unused and should not be used.

sophie-xhonneux committed:
Wrap latent state into a dataclass

to simply loss calculation later

clessig committed:
Removed unused parameters

Jubeku committed:
fixed trainer for multiple terms in losses_all, still need to fix logging

sophie-xhonneux committed:
Progress on computing the loss on correct dims

Added the new ViewMetadat and ModelBatch dataclasses that will come from
the cropping PR

Added LatentState dataclass to compute the latent heads on the correct
part of the latent state

TODOs:
1. Deal with DINO local and global component
2. Write JEPA loss function in loss.py
3. Test iBOT with actual mask and route student temperature
4. TODOs in the code

sophie-xhonneux committed:
Add views.py and run Ruff

clessig committed:
Inversion of target output ordering to match input one in forcast mode. Unclear how to deal with it with MTM

Jubeku committed:
fix _log_terminal

clessig committed:
Changes to  prepare_logging to apply index inversion

shmh40 committed:
added file with ModelBatch and SampleMetadata dataclasses

Jubeku commented on src/weathergen/train/loss_modules/loss_module_ssl.py:
```diff
@@ -0,0 +1,138 @@
+            )
+            loss_value = loss_fn(preds_for_loss, targets_for_loss).mean()
+            loss += weight * loss_value
+            losses_all[name] = loss_value.item()
```
Wait, `stream.name` anyways doesn't make sense for latent losses. Maybe it does make sense to allow name to be something general.

clessig committed:
Updating config to working version

Jubeku committed:
fix logging

shmh40 committed:
update ViewMetadata spec

shmh40 committed:
draft changes to allow global local view generation in masker and tokenizer_masking. generate the mask, otherwise using batchify_source and batchify_target as before, with the capacity to remember what mask we have now when it comes to generating the targets. Update to inputs_metadata structure but not put in to practice

shmh40 committed:
draft of training_config in default_config

shmh40 committed:
change view_metadata to dict in ModelInput

sophie-xhonneux committed:
Close in on completing DINO loss

TODO needs to deal with the tuple part of the DINO loss
TODO understand how the ModelBatch input structure affects the loss
terms

clessig committed:
NOT WORKING: updating class to handle multiple input steps and improving overall structure

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Added basic support for multi-step sources.

clessig committed:
Partially enabled correct handling of multiple input steps.

Jubeku committed:
initialize loss as torch tensor with grad

Jubeku committed:
remove level in hist losses dict

Jubeku committed:
rename loss.py to loss_functions.py

Jubeku committed:
rename loss.py to loss_functions.py

Jubeku committed:
return loss with grads seperately to trainer

clessig committed:
Added mode and refactored get_sample_data into separate function.

Jubeku committed:
modify log names

Jubeku committed:
add loss_functions.py

sophie-xhonneux committed:
Revert "rough first effort producing globaland local views"

This reverts commit 3fa00336e1b48a3bed5b703ef7c38bdc3a0acffd.

Jubeku committed:
merge develop

sophie-xhonneux committed:
Lint code

Jubeku committed:
rm loss_fcts in default config

clessig committed:
Comments

clessig committed:
Renaming

shmh40 committed:
updated default config training_config to allow student-teacher

shmh40 committed:
added stream id to era5 config

shmh40 committed:
slight restructure of ViewMetadata

shmh40 committed:
basic if statement to yield the student and teacher views

shmh40 committed:
merge changes

shmh40 committed:
correct imports with new batch.py

sophie-xhonneux committed:
Extract get batch size util function

Easier to read and as batchsize gets more complicated in SSL this will
be a useful abstraction

sophie-xhonneux committed:
Fix rebase of loss loss_calculator

shmh40 committed:
created function for _get_student_teacher_sample_data which returns the streams_data of the teacher and multiple streams_datas for the student views.

clessig committed:
Not working draft for restructuring

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Changes for better student teacher structure

clessig committed:
More refactoring

clessig committed:
More refactoring and cleanup

clessig committed:
More refactoring. Code working again.

clessig committed:
Cleaned up parametrization

clessig committed:
Changes necessary for spoofing flag per IOReaderData

clessig committed:
Changes to have spoofing on a per data reader sample

clessig committed:
Moved _get_student_teacher_masks() so that masks are generated for all streams first.

clessig committed:
Renaming and minor clean up.

clessig committed:
Added basic support for use of ModelBatch class to define rough structure and interface.

clessig committed:
linting

clessig committed:
Linting

clessig committed:
linting

clessig committed:
Linting problems but removed unused ViewMetaData dependence

clessig committed:
Added required reflexivity between source and target samples to Batch

clessig committed:
Added todo

sophie-xhonneux committed:
Test for compute time regressions

sophie-xhonneux committed:
Prepare for merge

sophie-xhonneux committed:
Lint the code

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/jk/develop/loss_calc_base' into sophiex/dev/abstract-class-teacher-1179

sophie-xhonneux committed:
Lint code

sophie-xhonneux committed:
Lint

Commit:
Fix some basic bugs

shmh40 committed:
fix typo in ModelBatch

shmh40 committed:
collect num_source_samples and num_target_samples, add loop over teacher masks hence allowing multiple teacher views, and add source_target_idx to keep track of which student belongs to which teacher

shmh40 committed:
add teacher num_views parameter to config

clessig committed:
Re-enabling inversion of targert ordering.

shmh40 committed:
tidy up, remove unused build_stream_views in tokenizer_masking

shmh40 committed:
multiple idxs for each teacher, need to confirm for not student case, and updated ModelBatch for this

tjhunter commented on src/weathergen/train/loss_modules/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .loss_module_latent import LossLatent
```
nit: would you mind using absolute paths? relative paths tend to break in mysterious and confusing ways

shmh40 committed:
add max_num_targets to era5

shmh40 committed:
add max_num_samples functionality to tokenizer_masking and pass through in multi_stream_data_sampler. coords_per_cell is a bit nasty

clessig committed:
Removing spurious code / things that should be merged later

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into sophiex/dev/abstract-class-teacher-1179

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into sophiex/dev/abstract-class-teacher-1179

clessig committed:
Linting

shmh40 committed:
move build_views_for_stream into masker

sophie-xhonneux committed:
Merge branch 'sophiex/dev/abstract-class-teacher-1179' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Lint code

sophie-xhonneux committed:
Rename identity TargetAndAux module

shmh40 committed:
tidy up, remove unused arguments, types

shmh40 committed:
fix masking for NPP-ATMS by correctly selecting final timestep mask and aligning between source and target. working for num_input_steps = 1, broken for > 1, compute_offsets_scatter_embed not working

sophie-xhonneux committed:
Make code runnable

shmh40 committed:
merge with develop, include trainer idx_inv_rt, merged default_config, rm tokenizer_forecast

shmh40 committed:
updated configs so code runs. Note default config to be overhauled still

clessig committed:
Draft for model interface

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

CAREFUL UNTESTED

TODO make runnable

sophie-xhonneux committed:
Make code runnable again

Seems slow again

clessig committed:
Cleaned up and restructured structure. Not working yet with FSDP

clessig committed:
Fixes for FSDP/DDP

clessig committed:
Cleaning up, should be merged when needed

clessig committed:
Fixes to FSDP

clessig committed:
Fix incorrect args for model loading and removing unused code.

clessig committed:
Linting

clessig committed:
Removing old code

clessig committed:
- Fixing inference arg order
- Fixing subtle problem with world_size_original that should be taken from config when available

clessig committed:
Fixing interface of get_target_aux_calculator

clessig committed:
Fixing call to target aux calculator

clessig committed:
Fixes to get_target_aux_calculator

sophie-xhonneux committed:
Remove stale dataclasses

clessig committed:
Fix MAE

shmh40 committed:
very hacky first pass of full masking_strategy_config for the student and teacher views. Much to fix up

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/clessig/dev/abstract-class-teacher-1179-model-interface' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

shmh40 committed:
instructions for sophie

shmh40 committed:
add SampleMetaData integration and functionality, and update masker to use SampleMetadata. Pass through source_cell_lens and target_coords_idx to student_teacher_batch in iter, and hence pass through to trainer. source_cell_lens and target_coords_idx are now part of Sample, which is itself the components of ModelBatch. To tidy

sophie-xhonneux committed:
Prepare for another merge

shmh40 committed:
remove prints, pdb

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Save state

shmh40 committed:
add mask to SampleMetaData and add forecast_dt to Sample so it is accessible. Can specify the loss in the default config with student-teacher views

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Save state for Seb

Currently re-viving the EMATeacher creation

Memory is an issue, had to hardcode a smaller latent space

shmh40 committed:
add diffusion forecast option for the data sampling, and with noise_level_rn in the metadata. The Trainer needs to be copied from Sophies branch, currently we only get so far

sophie-xhonneux committed:
Attemp to make the iBOT loss work

TODO force 1 ibot student view per global view
TODO there is a bug with the mask causing a leaf error in pytorch
TODO remove all the hardcoded reduced latent space

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Linting

clessig committed:
Simplified and clarified handling of default target_aux_calcualtor

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Restoring masking as training_mode in default_config

clessig committed:
More linting

clessig committed:
Removed duplicate lines due to mergeing

clessig committed:
Restored masking as training mode. Not working due to NaN in prediction

clessig committed:
Fixed problem in engines introduced in recent commits merging develop. This fixes masking training

shmh40 committed:
remove unused mask generation in diffusion_forecast

shmh40 committed:
restore masking_strategy to random

Had placeholder for testing, now back to "random" for masking strategy in the base level of default_config

shmh40 committed:
restore loader_num_workers to 8

shmh40 committed:
fix indentation of else: assert False in _get_sample msds

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

sophie-xhonneux committed:
Pipe data through all ssl loss fns

TODO iBOT head should output class tokens as well as patch tokens
TODO remove hardcoded assignments, should be based on config
TODO deal with the memory hungriness of it all
TODO carefully inspect for bugs

tjhunter committed:
merge with dev

tjhunter committed:
linter warnings

tjhunter committed:
commenting tests

clessig committed:
Restructured code so that mask generation and application is cleanly separated

clessig committed:
Commit

clessig committed:
Update

clessig committed:
Merge branch 'shmh40/dev/1270-idx-global-local' of github.com:ecmwf/WeatherGenerator into shmh40/dev/1270-idx-global-local

clessig committed:
Fixed uv.lock

clessig committed:
Fix for integration test

clessig committed:
Re-enabled multi-source training

Jubeku committed:
1390 - Adapt forward pass of new batch object (#1391)

* Add to device to ModelBatch, etc & adapt model

TODO adapt validate and inference
TODO test forecasting and multiple stream because predict changed
substantially

* Rename view to sample and fix validate

* Revert predict function and fix inference

* Fix invalid access with mask

* Linting

* Fixed handling of target_idxs and other minor issues

---------

Co-authored-by: sophiex <24638638+sophie-xhonneux@users.noreply.github.com>
Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Merge remote-tracking branch 'origin/shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

clessig committed:
Completed migration to new batch class by removing reference to old list of lists

clessig committed:
Fixed missing non_blocking=True in to_device()

clessig committed:
Removed old comments

clessig committed:
Fixed problem with non_blocking=True

clessig committed:
Cleaned up comments and return values a bit

clessig committed:
Changed args to embedding

clessig committed:
Changed core functions to take sample as arg

clessig committed:
Changed that model takes sample as input

clessig committed:
Fixes for diffusion

clessig committed:
Switched to lists of model / target stratgies

sophie-xhonneux committed:
Pipe the mask through

sophie-xhonneux committed:
Merge branch 'shmh40/dev/1270-idx-global-local' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Filter student views for the correct loss

sophie-xhonneux committed:
Change the masking and msdp to fit student-teacher

1. We ensure that for each target view all the student views are generated
2. We ensure that the target views have their mask applied to the input

sophie-xhonneux committed:
Make DINO and iBOT work

TODO: use the target mask to reduce memory

sophie-xhonneux committed:
Prepare for Model PR introducing class & reg token

Thus, right now it breaks. The big question is memory!

sophie-xhonneux committed:
Integrate the class and register token PR

Done manually because I couldn't figure out how to merge from a fork

sophie-xhonneux committed:
Fix iBOT loss with correct PredHead

Limitation: iBOT loss needs num_class_tokens to be 1

sophie-xhonneux committed:
Fix JEPA + Lint code

sophie-xhonneux committed:
Fix DDP

It had unused parameters from the decoders these had to be removed

sophie-xhonneux committed:
Running this code + config for JEPA with DDP

sophie-xhonneux committed:
Ran JEPA DDP plot with this

sophie-xhonneux committed:
Fix FSDP error

Potentially a slow down, but I don't understand FSDP well enough for a better fix

sophie-xhonneux committed:
Fix conig

kctezcan commented on config/default_config.yml:
```diff
@@ -91,39 +92,81 @@ validate_with_ema: True
+      # "JEPA": {'weight': 5, "loss_extra_args": {}, "out_dim": 2048} }
+# Student-teacher configuration (only used when training_mode == "student_teacher")
+# TODO: adapt so that the masking or forecast config entry also sits here
+training_config:
```
should we call these training_settings instead of _config to avoid having clashing names?

sophie-xhonneux replied:
which name would clash?

kctezcan commented on src/weathergen/datasets/stream_data.py:
```diff
@@ -131,7 +129,6 @@ def add_empty_target(self, fstep: int) -> None:
 
         self.target_tokens[fstep] += [torch.tensor([], dtype=torch.int32)]
-        self.target_tokens_lens[fstep] += [torch.zeros([self.healpix_cells], dtype=torch.int32)]
         self.target_coords[fstep] += [torch.zeros((0, 105)) for _ in range(self.healpix_cells)]
```
sure about the hardcoded 105 here?

clessig replied:
The 105 is there for a long time. But maybe someone can tidy this up ;)

sophie-xhonneux committed:
Merge branch 'develop' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Fix validation

sophie-xhonneux committed:
Stuck on error taking a break

shmh40 committed:
hot fix to empty tokens_c in encoder when looping over chunks

shmh40 committed:
Revert "hot fix to empty tokens_c in encoder when looping over chunks"

This reverts commit e4519d8b910b15470d21cecc0ec71062710dd4a6.

shmh40 committed:
hot fix for local assimilation empty tokens_c

sophie-xhonneux committed:
Add class tokens being variable + Fix bugs

sophie-xhonneux committed:
Push remaining changes to default config

shmh40 committed:
deepcopy configs so we do not pop weight and lose it for inference

shmh40 committed:
fixed bug in inference with +2 in forecast steps range

shmh40 committed:
add required import to trainer

sophie-xhonneux committed:
Merge branch 'develop' into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Update uv.lock

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into sophiex/dev/ssl-losses-1043

clessig committed:
Linting

sophie-xhonneux committed:
Record fstep latent states

shmh40 committed:
added two configs, jepa and ibot/dino. Note these configs still try to merge/overwrite the default config with the --config flag

sophie-xhonneux commented on src/weathergen/train/ssl_losses_utils.py:
```diff
@@ -0,0 +1,304 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
You want to rename the file simply? I can do that, just want to make sure I understand correctly

clessig replied:
Yes, I would rename it, and move to weathergen/models/. Sorry, the comment was incomplete. For me, there is key functionality in this file and I wouldn't expect it in a file called utils. For me it's also mainly a model part although I see that it's close to the loss computation.

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Addres comments from PR review

sophie-xhonneux committed:
Prepare SSL losses for logging

Currently nothing happens in the terminal but I don't know why that is

clessig committed:
Merge branch 'sophiex/dev/ssl-losses-1043' of github.com:ecmwf/WeatherGenerator into sophiex/dev/ssl-losses-1043

sophie-xhonneux committed:
Lint

sophie-xhonneux committed:
Address PR comments+ upstream changes

sophie-xhonneux committed:
Appease the hidden linter

sophie-xhonneux committed:
Rename ssl_losses_utils

sophie-xhonneux committed:
Add the untracked file

clessig committed:
Removing spurious character

clessig reviewed (approved):
Awesome work! Excited to see the experiments!
